### PR TITLE
Improved SRS Design with args error checking, better types

### PR DESCRIPTION
**Pull request by smishr**

### CODE REVIEW

**Comment by ayushpatnaikgit** [comment]
@smishr can you check where the test is failing?

File: src/SurveyDesign.jl [hunk 1]
```diff
@@ -2,13 +2,13 @@
 function print_short(x)
     # write floats in short form
     if isa(x[1], Float64)
-        x = round.(x, sigdigits = 3)
+        x = round.(x, sigdigits=3)
```
Comment (iuliadmtru) [inline src/SurveyDesign.jl:5]: I guess this is the formatter's doing. In that case, let's keep the style like this - without spaces around equal signs for argument values - _but_ this should be a different PR. Also, this needs to be done in all the files in order to keep consistency.

File: src/SurveyDesign.jl [hunk 2]
```diff
@@ -30,53 +29,93 @@ The population size is equal to the sample size unless `popsize` is explicitly p
 """
 struct SimpleRandomSample <: AbstractSurveyDesign
     data::DataFrame
-    sampsize::UInt
-    popsize::Union{UInt,Nothing}
+    sampsize::Union{Nothing,Unsigned}
```
Comment (iuliadmtru) [inline src/SurveyDesign.jl:32]: Why is this change necessary?

File: src/SurveyDesign.jl [hunk 3]
```diff
@@ -30,53 +29,93 @@ The population size is equal to the sample size unless `popsize` is explicitly p
...
-        if isa(weights, Symbol)
-            weights = data[!, weights]
-        end
-        # set population size if it is not given; `weights` and `sampsize` must be given
+        popsize=nothing,
+        sampsize=nrow(data),
+        weights=nothing, # Check the defaults
+        probs=nothing,
+        ignorefpc=false
+    )
```
Comment (iuliadmtru) [inline src/SurveyDesign.jl:43]: Maybe this parenthesis should be removed? I don't see where it is coming from.

File: src/SurveyDesign.jl [hunk 4]
```diff
@@ -30,53 +29,93 @@ The population size is equal to the sample size unless `popsize` is explicitly p
...
+            equal_weight = first(weights)
+            popsize = round(sampsize .* equal_weight) |> UInt
+        elseif typeof(popsize) <: Vector{<:Real}
+            if !all(y -> y == first(popsize), popsize) # SRS by definition is equi-weighted
+                error("Simple Random Sample must be equi-weighted. Different sampling weights detected in vectors")
+            end
+            weights = popsize ./ sampsize
+            popsize = first(popsize) |> UInt
+        else
+            error("If popsize not given then either sampling weights or sampling probabilities must be given")
```
Comment (iuliadmtru) [inline src/SurveyDesign.jl:78]: "If popsize **is** not given ..."

File: src/SurveyDesign.jl [hunk 5]
```diff
@@ -30,53 +29,93 @@ The population size is equal to the sample size unless `popsize` is explicitly p
...
-        fpc = ignorefpc ? 1 : 1 - (sampsize / popsize)
-
+        fpc = ignorefpc ? 1 : 1 .- (sampsize ./ popsize)
+        # Add columns for weights and probs in data -- Check if really needed to add them as columns
+        if !isnothing(probs)
+            # add probability weights column to `data`
+            data[!, :probs] = probs
+            # add frequency weights column to `data`
+            data[!, :weights] = 1 ./ data[!, :probs]
+        else # else weights were specified
```
Comment (iuliadmtru) [inline src/SurveyDesign.jl:90]: Remove "else" from the comment

File: src/SurveyDesign.jl [hunk 6]
```diff
@@ -30,53 +29,93 @@ The population size is equal to the sample size unless `popsize` is explicitly p
...
+    printstyled("data: "; bold=true)
     print(size(design.data, 1), "x", size(design.data, 2), " DataFrame")
-    printstyled("\nweights: "; bold = true)
+    printstyled("\nweights: "; bold=true)
     print_short(design.data.weights)
-    printstyled("\nprobs: "; bold = true)
+    printstyled("\nprobs: "; bold=true)
     print_short(design.data.probs)
-    printstyled("\nfpc: "; bold = true)
+    printstyled("\nfpc: "; bold=true)
```
Comment (iuliadmtru) [inline src/SurveyDesign.jl:109]: Maybe have a check here. 

```julia
printstyled("\nignore fpc: "; bold=true)
print(design.ignorefpc)

if (!design.ignorefpc)
    printstyled("\nfpc: "; bold=true)
    print_short(design.fpc)
end
```

File: src/SurveyDesign.jl (hunk 1 above)
Comment (smishr) [reply src/SurveyDesign.jl:5]: Yep the formatter reduces the space around `=` signs

File: src/SurveyDesign.jl (hunk 2 above)
Comment (smishr) [reply src/SurveyDesign.jl:32]: Unsigned is the `supertype` for UInt, more correct

File: src/SurveyDesign.jl (hunk 6 above)
Comment (smishr) [reply src/SurveyDesign.jl:109]: No, if ignorefpc is true, then fpc should print as `1`, just as sanity check

**Comment by ayushpatnaikgit** [comment]
Shall we merge this?

**Comment by smishr** [comment]
> Shall we merge this?

yes
